File: ruby_small_problems/easy_4/8.rb
# takes a string of numeric chars including an optional plus or minus sign
# returns an integer
DIGITS = {
  '0' => 0, '1' => 1, '2' => 2, '3' => 3, '4' => 4,
  '5' => 5, '6' => 6, '7' => 7, '8' => 8, '9' => 9
}

def string_to_integer(str)
  digits = str.chars.map{ |x| DIGITS[x]  }
  value = 0
  digits.each { |x| value = (10 * value) + digit }
  value
end

# jaron truman's is sleek
def string_to_signed_integer(str)
  unsigned = str.start_with?('+', '-') ? str[1..-1] : str
  number = string_to_integer(unsigned)
  str[0] == '-' ? -number : number
end

File: 210/exercises/javascript_language_fundamentals/javascript_basics/8.js
const DIGITS = {
  0: 0,
  1: 1,
  2: 2,
  3: 3,
  4: 4,
  5: 5,
  6: 6,
  7: 7,
  8: 8,
  9: 9,
}

function stringToSignedInteger(string) {
  let multiplier = 1;
  let i;
  let result = 0;
  let start = string.length - 1;
  let end = 0;

  if (string[0] == '-' || string[0] == '+') {
    end += 1;
  }

  for (i = string.length - 1; i >= end; i -= 1) {
    result += DIGITS[string[i]] * multiplier; 
    multiplier *= 10;
  }

  if (string[0] == '-') {
    result = 0 - result; 
  }

  return result;
}

console.log(stringToSignedInteger('4321'));      // 4321
console.log(stringToSignedInteger('-570'));      // -570
console.log(stringToSignedInteger('+100'));      // 100
